### PR TITLE
Implementation of stack creation

### PR DESCRIPTION
**Pull request by devosray**
This is the first revision of the implementation of the core stack creation APIs.
To run the example cURL scripts, you first need to set a helper environmental variable to the value of your JWT token:

```$ TOKEN=eyJhbGciO... ```

- GET /stacks
```
~ curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X GET 127.0.0.1:8080/stacks
=> {"data":[{"stackName":"docks-api","servicesCount":2},{"stackName":"more-tests32","servicesCount":1},{"stackName":"nginx-lol","servicesCount":4}]}
```
- PUT /stacks
```
curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X PUT 127.0.0.1:8080/stacks/demo-stack -d '{"stackFile": "dmVyc2lvbjogJzMnCnNlcnZpY2VzOgogICAgd2ViOgogICAgICAgIGltYWdlOiBuZ2lueAogICAgICAgIHBvcnRzOgogICAgICAgICAgICAtICI2OTY5OjgwIgo="}'
=> Updating service demo-stack_web (id: uapmh7zc2fi9oo9aj25k94pyw)
```

- POST /stacks
```
~ curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X POST 127.0.0.1:8080/stacks -d '{"stackName": "demo-stack", "stackFile": "dmVyc2lvbjogJzMnCnNlcnZpY2VzOgogICAgd2ViOgogICAgICAgIGltYWdlOiBuZ2lueAogICAgICAgIHBvcnRzOgogICAgICAgICAgICAtICI2OTY5OjgwIgo="}'
=> Creating service demo-stack_web
```

- GET /stacks/[stackname]/services
```
curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X GET 127.0.0.1:8080/stacks/demo-stack/services           
=> {"data":[{"ID":"uapmh7zc2fi9oo9aj25k94pyw","Version":{"Index":218},"CreatedAt":"2018-07-19T10:24:17.839501028Z","UpdatedAt":"2018-07-19T10:25:00.056166227Z","Spec":{"Name":"demo-stack_web","Labels":{"com.docker.stack.namespace":"demo-stack"},"TaskTemplate":{"ContainerSpec":{"Image":"nginx","Labels":{"com.docker.stack.namespace":"demo-stack"},"Isolation":"default"},"Resources":{},"Placement":{},"Networks":[{"Target":"usl8oi341td06pri7b8avwkvc","Aliases":["web"]}],"ForceUpdate":0,"Runtime":"container"},"Mode":{"Replicated":{"Replicas":1}},"EndpointSpec":{"Mode":"vip","Ports":[{"Protocol":"tcp","TargetPort":80,"PublishedPort":6969,"PublishMode":"ingress"}]}},"PreviousSpec":{"Name":"demo-stack_web","Labels":{"com.docker.stack.namespace":"demo-stack"},"TaskTemplate":{"ContainerSpec":{"Image":"nginx","Labels":{"com.docker.stack.namespace":"demo-stack"},"Isolation":"default"},"Resources":{},"Placement":{},"Networks":[{"Target":"usl8oi341td06pri7b8avwkvc","Aliases":["web"]}],"ForceUpdate":0,"Runtime":"container"},"Mode":{"Replicated":{"Replicas":1}},"EndpointSpec":{"Mode":"vip","Ports":[{"Protocol":"tcp","TargetPort":80,"PublishedPort":6969,"PublishMode":"ingress"}]}},"Endpoint":{"Spec":{"Mode":"vip","Ports":[{"Protocol":"tcp","TargetPort":80,"PublishedPort":6969,"PublishMode":"ingress"}]},"Ports":[{"Protocol":"tcp","TargetPort":80,"PublishedPort":6969,"PublishMode":"ingress"}],"VirtualIPs":[{"NetworkID":"ilzl91dq82cjf7s1rt8ulh2dy","Addr":"10.255.0.17/16"},{"NetworkID":"usl8oi341td06pri7b8avwkvc","Addr":"10.0.3.3/24"}]}}]}
```

- GET /stacks/[stackname]/tasks
```
curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X GET 127.0.0.1:8080/stacks/demo-stack/tasks   
= > {"data":[{"ID":"tunmpq4ia02w2bygbgtqj81vw","Version":{"Index":217},"CreatedAt":"2018-07-19T10:24:17.858015018Z","UpdatedAt":"2018-07-19T10:24:29.960363376Z","Labels":{},"Spec":{"ContainerSpec":{"Image":"nginx","Labels":{"com.docker.stack.namespace":"demo-stack"},"Isolation":"default"},"Resources":{},"Placement":{},"Networks":[{"Target":"usl8oi341td06pri7b8avwkvc","Aliases":["web"]}],"ForceUpdate":0},"ServiceID":"uapmh7zc2fi9oo9aj25k94pyw","Slot":1,"NodeID":"y1tcqv87uwyylw1suezys4v8z","Status":{"Timestamp":"2018-07-19T10:24:29.938504838Z","State":"running","Message":"started","ContainerStatus":{"ContainerID":"dc2c71fdb08d0701c144ec7e903c02598570340d90c8ee0208628c0ef149571e","PID":14118,"ExitCode":0},"PortStatus":{}},"DesiredState":"running","NetworksAttachments":[{"Network":{"ID":"ilzl91dq82cjf7s1rt8ulh2dy","Version":{"Index":135},"CreatedAt":"2018-07-18T07:12:33.583465495Z","UpdatedAt":"2018-07-19T10:00:18.798069635Z","Spec":{"Name":"ingress","Labels":{},"DriverConfiguration":{},"Ingress":true,"IPAMOptions":{"Driver":{},"Configs":[{"Subnet":"10.255.0.0/16","Gateway":"10.255.0.1"}]},"Scope":"swarm"},"DriverState":{"Name":"overlay","Options":{"com.docker.network.driver.overlay.vxlanid_list":"4096"}},"IPAMOptions":{"Driver":{"Name":"default"},"Configs":[{"Subnet":"10.255.0.0/16","Gateway":"10.255.0.1"}]}},"Addresses":["10.255.0.18/16"]},{"Network":{"ID":"usl8oi341td06pri7b8avwkvc","Version":{"Index":198},"CreatedAt":"2018-07-19T10:23:15.164446118Z","UpdatedAt":"2018-07-19T10:23:15.188455925Z","Spec":{"Name":"demo-stack_default","Labels":{"com.docker.stack.namespace":"demo-stack"},"DriverConfiguration":{"Name":"overlay"},"Scope":"swarm"},"DriverState":{"Name":"overlay","Options":{"com.docker.network.driver.overlay.vxlanid_list":"4100"}},"IPAMOptions":{"Driver":{"Name":"default"},"Configs":[{"Subnet":"10.0.3.0/24","Gateway":"10.0.3.1"}]}},"Addresses":["10.0.3.4/24"]}]}]}
```
- DELETE /stacks/[stackname]
```
~ curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X DELETE 127.0.0.1:8080/stacks/demo-stack
=>
```

### CODE REVIEW

**Comment by egeldenhuys** [comment]
The following request is hanging
```
curl -X "GET" -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImFkbWluIiwiaWF0IjoxNTMxNzYwNDQwfQ.a3dbFn7SuAiMPFyDtKLTz6Zh7EIKYqqaDpExvfUdjRQ" http://127.0.0.1:8080/stacks
```

Am I doing anything wrong?

**Comment by devosray** [comment]
@egeldenhuys I suspect the issue is with the production image not having the Docker client. I added it to the Dockerfile and Docker Cloud rebuilt it. Could you pull the new image and test if the problem persists?

**Comment by devosray** [comment]
I have updated my pull request to include some example cURL calls to the API. I believe this is ready to be merged into develop.

File: routes/stacksApi.js [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+  const stackList = await retrieveStackList();
+  res.status(200).send({
+    data: stackList,
+  });
+});
+
+// Deploy a new stack to the Swarm. The stack name should not exist
+router.post('/', async function(req, res, next) {
+  if (!req.body.hasOwnProperty('stackName') || req.body['stackName'] === '') {
+    res.status(400).send('Required parameter stackName missing');
```
Comment (egeldenhuys) [inline routes/stacksApi.js]: Good use of HTTP 400

**Comment by egeldenhuys** [comment]
Some unexpected behaviour that makes frontend logic less logical. This is more documenting Docker's features than the API. The only thing missing from the API response are the lines where Docker actually creates a service before failing.

**Stackfile**
```
version: "3"

services:
  service1:
    image: nginx
    ports:
      - 54001:80

  service2:
    image: nginx
```

1. Set the base64 stack file
```
$ export STACKFILE=dmVyc2lvbjogIjMiCgpzZXJ2aWNlczoKICBzZXJ2aWNlMToKICAgIGltYWdlOiBuZ2lueAogICAgcG9ydHM6CiAgICAgIC0gNTQwMDE6ODAKCiAgc2VydmljZTI6CiAgICBpbWFnZTogbmdpbngKCiAgICAK
```

2. Deploy the stack. Everything should go fine
```
$ curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X POST 127.0.0.1:8080/stacks -d "{\"stackName\": \"lolwat1\", \"stackFile\": \"$STACKFILE\"}" -D -
HTTP/1.1 200 OK
X-Powered-By: Express
Access-Control-Allow-Origin: *
Content-Type: text/html; charset=utf-8
Content-Length: 101
ETag: W/"65-E0m6DFQaUdPcol3VN2mmEmjUGa0"
Date: Thu, 19 Jul 2018 16:52:50 GMT
Connection: keep-alive

Creating network lolwat1_default
Creating service lolwat1_service1
Creating service lolwat1_service2
```

2. Deploy the same stack again. Should get an error on the port being in use
```
curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X POST 127.0.0.1:8080/stacks -d "{\"stackName\": \"lolwat2\", \"stackFile\": \"$STACKFILE\"}" -D -
HTTP/1.1 500 Internal Server Error
X-Powered-By: Express
Access-Control-Allow-Origin: *
Content-Type: text/html; charset=utf-8
Content-Length: 177
ETag: W/"b1-g5XskqaQTEiDO/lHPdgnKT3pRcY"
Date: Thu, 19 Jul 2018 16:52:57 GMT
Connection: keep-alive

Error response from daemon: rpc error: code = InvalidArgument desc = port '54001' is already in use by service 'lolwat1_service1' (cmttbxq4tn6te5n61rbojw4m1) as an ingress port
```

3. Deploy the same stack again. Should get an error on the port being in use
```
curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X POST 127.0.0.1:8080/stacks -d "{\"stackName\": \"lolwat3\", \"stackFile\": \"$STACKFILE\"}" -D -
HTTP/1.1 500 Internal Server Error
X-Powered-By: Express
Access-Control-Allow-Origin: *
Content-Type: text/html; charset=utf-8
Content-Length: 177
ETag: W/"b1-g5XskqaQTEiDO/lHPdgnKT3pRcY"
Date: Thu, 19 Jul 2018 16:53:09 GMT
Connection: keep-alive

Error response from daemon: rpc error: code = InvalidArgument desc = port '54001' is already in use by service 'lolwat1_service1' (cmttbxq4tn6te5n61rbojw4m1) as an ingress port
```

4. Verify Docker state. lolwat3 exists but lolwat2 does not. Docker appears to deploy the services in a stack in a round robin like fashion. The second time it fails on the first service. The third time it starts with the second service and successfully deploys it, but fails with the first service.
```
$ docker stack ls
NAME                SERVICES
lolwat1             2
lolwat3             1

$ docker network ls
NETWORK ID          NAME                              DRIVER              SCOPE
qx3nrgjq7pr0        lolwat1_default                   overlay             swarm
u9hgyoacck62        lolwat2_default                   overlay             swarm
3447ki1fuifg        lolwat3_default                   overlay             swarm
```

5. Try with CLI
```
$ docker stack deploy -c bad.yml cli1
Creating network cli1_default
Creating service cli1_service1
failed to create service cli1_service1: Error response from daemon: rpc error: code = InvalidArgument desc = port '54001' is already in use by service 'lolwat1_service1' (cmttbxq4tn6te5n61rbojw4m1) as an ingress port
```
### Conclusions from Docker CLI
Once a request to deploy a stack has been made, assume some parts have been deployed.

### Comments on Docks API
Is it possible to also capture and return the following to give the user more context?
```
Creating network cli1_default
Creating service cli1_service1
```

**Comment by devosray** [comment]
Ah, good spot. Thanks for the solid write-up. I'll fix it to not fail-fast but also return the standard output, as in your example.

**Comment by devosray** [comment]
I have modified the response accordingly:


```
$ curl -H "Authorization: Bearer $TOKEN" -H "Accept: application/json" -H "Content-type: application/json" -X POST 127.0.0.1:8080/stacks -d "{\"stackN
ame\": \"lolwat2\", \"stackFile\": \"$STACKFILE\"}" -D -                                                                                                                                             
                                                                                                                                                                                                     
HTTP/1.1 500 Internal Server Error                                                                                                                                                                   
X-Powered-By: Express                                                                                                                                                                                
Access-Control-Allow-Origin: *                                                                                                                                                                       
Content-Type: text/html; charset=utf-8                                                                                                                                                               
Content-Length: 212                                                                                                                                                                                  
ETag: W/"d4-MvymCi8UFleEfu2oAbguePXK+9Y"                                                                                                                                                             
Date: Thu, 19 Jul 2018 17:37:29 GMT                                                                                                                                                                  
Connection: keep-alive                                                                                                                                                                               
                                                                                                                                                                                                     
Creating service lolwat2_service1                                                                                                                                                                    
                                                                                                                                                                                                     
Error response from daemon: rpc error: code = InvalidArgument desc = port '54001' is already in use by service 'lolwat1_service1' (7t7mgtme8eyqgknj2o6ic5vpw) as an ingress port
```

**Comment by egeldenhuys** [comment]
Thank you for that

**Pull request merged**